share/steel/examples/pulse/lib/Pulse.Lib.HigherArray.fst

```diff
@@ -0,0 +1,875 @@
+    (mk_carrier (SZ.v (ptr_of arr).base_len) (ptr_of arr).offset s (half_perm p));
+  fold pts_to arr #(half_perm p) s;
+  fold pts_to arr #(half_perm p) s;
+}
+```
+let share = share'
+
+let mk_carrier_gather
+  (#elt: Type)
+  (len: nat)
+  (offset: nat)
+  (s1 s2: Seq.seq elt)
+  (p1 p2: perm)
+  (_:squash (
+    let c1 = mk_carrier len offset s1 p1 in
+    let c2 = mk_carrier len offset s2 p2 in
+    composable c1 c2 /\
+    Seq.length s1 == Seq.length s2 /\
+    offset + Seq.length s1 <= len
+  ))
+: squash (
+    let c1 = mk_carrier len offset s1 p1 in
+    let c2 = mk_carrier len offset s2 p2 in
+      composable c1 c2 /\
+      mk_carrier len offset s1 (p1 `sum_perm` p2) == (c1 `compose` c2) /\
+      mk_carrier len offset s2 (p1 `sum_perm` p2) == (c1 `compose` c2) /\
+      s1 == s2
+  )
+=
+  let c1 = mk_carrier len offset s1 p1 in
+  let c2 = mk_carrier len offset s2 p2 in
+  assert (composable c1 c2);
+  assert (mk_carrier len offset s1 (p1 `sum_perm` p2) `Map.equal` (c1 `compose` c2));
+  assert (mk_carrier len offset s2 (p1 `sum_perm` p2) `Map.equal` (c1 `compose` c2));
+  mk_carrier_inj len offset s1 s2 (p1 `sum_perm` p2) (p1 `sum_perm` p2)
+
+
+let mk_carrier_valid_sum_perm
+  (#elt: Type)
+  (len: nat)
+  (offset: nat)
+  (s: Seq.seq elt)
+  (p1 p2: perm)
+: squash
+  (let c1 = mk_carrier len offset s p1 in
+   let c2 = mk_carrier len offset s p2 in
+   composable c1 c2 <==> valid_sum_perm len offset (Seq.length s) p1 p2)
+= let c1 = mk_carrier len offset s p1 in
+  let c2 = mk_carrier len offset s p2 in
+  if Seq.length s > 0 && offset + Seq.length s <= len
+  then
+    let open FStar.Real in
+    assert (Frac.composable (Map.sel c1 offset) (Map.sel c2 offset) <==> valid_perm len offset (Seq.length s) (sum_perm p1 p2))
+  else ()
+
+```pulse
+ghost
+fn of_squash (#p:prop) (s:squash p)
+requires emp
+ensures pure p
+{
+  ()
+}
+```
+
+```pulse
+ghost
+fn gather'
+  (#a:Type)
+  (arr:array a)
+  (#s0 #s1:Ghost.erased (Seq.seq a))
+  (#p0 #p1:perm)
+requires pts_to arr #p0 s0 ** pts_to arr #p1 s1
+ensures pts_to arr #(sum_perm p0 p1) s0 ** pure (s0 == s1)
+{
+  unfold pts_to arr #p0 s0;
+  with w0. assert (pcm_pts_to (ptr_of arr).base w0);
+  unfold pts_to arr #p1 s1;
+  with w1. assert (pcm_pts_to (ptr_of arr).base w1);
+  Pulse.Lib.Core.gather (ptr_of arr).base w0 w1;
+  of_squash (mk_carrier_gather (SZ.v (ptr_of arr).base_len) ((ptr_of arr).offset) s0 s1 p0 p1 ());
+  of_squash (mk_carrier_valid_sum_perm (SZ.v (ptr_of arr).base_len) ((ptr_of arr).offset) s0 p0 p1);
+  fold pts_to arr #(sum_perm p0 p1) s0;
+}
+```
+let gather = gather'
+
+let ptr_shift
+  (#elt: Type)
+  (p: ptr elt)
+  (off: nat {offset p + off <= base_len (base p)})
+: ptr elt
+= {
+    base_len = p.base_len;
+    base = p.base;
+    offset = p.offset + off;
+  }
+
+let split_l'
+    (#elt: Type)
+    (a: array elt)
+    (i: erased nat {i <= length a})
+: array elt
+= { p = ptr_of a; length=i }
+
+irreducible
```

Comment:
It was important to mark these irreducible otherwise Pulse's calls to the F* unifier take extremely long, seeming to on some occasions even loop for minutes! I should distill that behavior into a standalone test.